### DISCUSSION TITLE
Issues with S3 using own S3 bucket in lambda deployment

### DISCUSSION THREAD

**Discussion by f-skold** [developmentseed/titiler#479] [General]
I have tried to use titiler / titiler-pds to display tiles for Sentinel 2.
Since I aim at using this at a larger scale, and we have the rest of the soultion on AWS in eu-north-1 ("Stockholm").
So I am aiming at using an S3 bucket from my account (after issues with private bucket, I changed to a public bucket with requester pays).

I liked the CDK code for titiler better than what was in titiler-pds, so I thought I would use that repository,

Wanting to run custom code forced me to rewrite the Dockerfile used for the lambda deployment to use the current code instead of already delivered code.

I use a modified verison of https://github.com/cogeotiff/rio-tiler-pds/blob/master/rio_tiler_pds/sentinel/aws/sentinel2.py
to control S3 Bucket name (and possibly path used).

When using my **public** requester pays-bucket, I got this message in log from executing the request.

I am trying to access the info endpoint.
` wget https://MYSTACK.execute-api.eu-north-1.amazonaws.com/scenes/sentinel/info?sceneid="S2A_MSIL2A_20220210T102151_N0400_R065_T33VWF_20220210T141713"`

And this is what I get in the logs:
```
OpenBLAS WARNING - could not determine the L2 cache size on this system, assuming 256k

START RequestId: 424c0bd2-20e7-44b8-a8f5-390064ab7490 Version: $LATEST

[WARNING]	2022-05-18T16:51:12.446Z	424c0bd2-20e7-44b8-a8f5-390064ab7490	CPLE_AppDefined in HTTP response code on https://BUCKETNAME.s3.amazonaws.com/sentinel-s2-l2a-cogs/33/V/WF/2022/2/S2A_33VWF_20220210_0_L2A/B03.tif: 0

[WARNING]	2022-05-18T16:51:12.447Z	424c0bd2-20e7-44b8-a8f5-390064ab7490	CPLE_AppDefined in HTTP response code on https://BUCKETNAME.s3.amazonaws.com/sentinel-s2-l2a-cogs/33/V/WF/2022/2/S2A_33VWF_20220210_0_L2A/B08.tif: 0

END RequestId: 424c0bd2-20e7-44b8-a8f5-390064ab7490

REPORT RequestId: 424c0bd2-20e7-44b8-a8f5-390064ab7490	Duration: 738.03 ms	Billed Duration: 739 ms	Memory Size: 3008 MB	Max Memory Used: 274 MB	Init Duration: 3774.27 ms
```

Running the same code locally in a docker image worked, I could point a local QGis client towards titiler, probably using WMTSCapabilities.xml,

From this, I deduce that I am able to read the tileInfo.json using the boto-wrapper:

`self.tileInfo = json.loads(get_object(self.hostname, f"{prefix2}/tileInfo.json"))`
I know this since I have try -expect around this statement and I see nothing about this. So I assume the first boto-call to getting an object works fine.

But reading the actual cogeo-Tif (the bands) fails:

Do I need to enable S3 endpoints for the /vsis3-drivers to work?

I looked at the code in S2L2ACOGReader, and thought that the URL to the file in the S3 might be wrong,
so I tried to modify `_get_band_url()`, and fill in what looks like a more correct URL (I have no idea what vsis3 does)

   bucket_hostname = self.hostname + ".s3.eu-north-1.amazonaws.com";
   return f"{self._scheme}://{bucket_hostname}/{prefix}/{band}.tif"

WIth this code, using a differnt URL, I got a different error message:

    CPLReleaseMutex: Error = 1 (Operation not permitted)

I did not manage to get anything conclusive from this last error message.

Otherwise, any tips on how to debug things in the lambda?

**Comment by f-skold** [comment]
Foun this: https://gdal.org/api/cpl.html#_CPPv415VSIGetSignedURLPKc12CSLConstList

> For example “/vsis3/bucket/filename” will be expanded as “https://bucket.s3.amazon.com/filename?X-Amz-Algorithm=AWS4-HMAC-SHA256…”

So it seems like it might have been wrong to have a separate "bucket_hostname" in my previous entry. (No idea how we get from s3://bucket/path/file to /vsis3/bucket/path/file.

So this sets me back to the first step.

**Reply by vincentsarago** [reply]
> (No idea how we get from s3://bucket/path/file to /vsis3/bucket/path/file.

rasterio will translate `s3://` url to `vsis3` url for you 

> CPLReleaseMutex: Error = 1 (Operation not permitted)

make sure to not use Threading (set `RIO_TILER_MAX_THREADS=1`) in your env

I'll try to answer the other question tomorrow.

**Comment by f-skold** [comment]
Now I got nothing usefull at all in the AWS Cloudwatch logs, but I still get http code 500, internal server error.

wget output, (Hostname and IP numbers modified.)

```
--2022-05-24 11:59:06--  https://MYSTACK.execute-api.eu-north-1.amazonaws.com/scenes/sentinel/info?sceneid=S2A_MSIL2A_20220210T102151_N0400_R065_T33VWF_20220210T141713
Resolving MYSTACK.execute-api.eu-north-1.amazonaws.com (MYSTACK.execute-api.eu-north-1.amazonaws.com)... 16.123.123.123, 13.234.234.234
Connecting to MYSTACK.execute-api.eu-north-1.amazonaws.com (MYSTACK.execute-api.eu-north-1.amazonaws.com)|16.123.123.123|:443... connected.
HTTP request sent, awaiting response... 500 Internal Server Error
2022-05-24 11:59:08 ERROR 500: Internal Server Error.
```

**Comment by f-skold** [comment]
Now this command has started working, (and the info endpoint as well) 

```
wget https://MYSTACK.execute-api.eu-north-1.amazonaws.com/scenes/sentinel/statistics?sceneid="S2A_MSIL2A_20220210T102151_N0400_R065_T33VWF_20220210T141713"
```

I have found several mistakes on my side so far.

+ I used the wrong name of the json metadata file. I used tileInfo.json like the S3 bucket with jpeg2000 files.
+ I had removed the "requst_pays=True" to `get_object()`
+ More mistakes on my part, mixing up the jpeg2000 and coGeoTif code

**Problems in current code/documentation:**

Found after consulting https://developmentseed.org/titiler/deployment/aws/intro/: 
+ `deployment/aws/.env.example` contains **TITILER_STACK_ADDITIONAL_ENV** instead of TITILER_STACK_ENV
